supreme/src/commonMain/kotlin/at/asitplus/signum/ecmath/RFC9380.kt

```diff
@@ -0,0 +1,219 @@
+package at.asitplus.signum.ecmath
+
+import at.asitplus.signum.indispensable.Digest
+import at.asitplus.signum.indispensable.ECCurve
+import at.asitplus.signum.indispensable.ECPoint
+import at.asitplus.signum.indispensable.misc.BitLength
+import at.asitplus.signum.supreme.hash.digest
+import com.ionspin.kotlin.bignum.integer.BigInteger
+import com.ionspin.kotlin.bignum.integer.Sign
+import com.ionspin.kotlin.bignum.modular.ModularBigInteger
+import kotlinx.io.Buffer
+import kotlinx.io.readByteArray
+import org.kotlincrypto.SecureRandom
+import kotlin.experimental.xor
+import kotlin.jvm.JvmInline
+
+private typealias hash_to_field = (Sequence<ByteArray>, Int) -> Array<ModularBigInteger>
+private typealias map_to_curve = (ModularBigInteger) -> ECPoint
+private typealias clear_cofactor = (ECPoint) -> ECPoint
+
+/** RFC 9380 8.2ff */
+private inline val ECCurve.Z get() = BigInteger(when (this) {
+    ECCurve.SECP_256_R_1 -> -10
+    ECCurve.SECP_384_R_1 -> -12
+    ECCurve.SECP_521_R_1 -> -4
+}).toModularBigInteger(this.modulus)
+
+private inline fun ModularBigInteger.sqrt() =
+    pow(modulus.plus(1).div(4))
+
+private inline val ECCurve.c1 get() = this.modulus.minus(3).div(4)
+private inline val ECCurve.c2 get() = (-Z).sqrt()
+
+/** log(modulus) * (3/2), this matches RFC9380 NIST curve suites, and also matches RFC9497 */
+private inline val ECCurve.L get() = when(this) {
+    ECCurve.SECP_256_R_1 -> 48
+    ECCurve.SECP_384_R_1 -> 72
+    ECCurve.SECP_521_R_1 -> 98
+}
+
+/** per RFC9794 4.7.2. */
+fun ECCurve.randomScalar() = SecureRandom().nextBytesOf(this.L).let { BigInteger.fromByteArray(it, Sign.POSITIVE).toModularBigInteger(this.order) }
+
+private inline fun clear_cofactor_trivial(p: ECPoint) = p.curve.cofactor * p
+
+private typealias expand_message = (msg: Sequence<ByteArray>, domain: ByteArray, lenInBytes: Int) -> Buffer
+
+private infix fun ByteArray.xor(other: ByteArray): ByteArray {
+    check(this.size == other.size)
+    return ByteArray(this.size) { i -> this[i] xor other[i] }
+}
+
+private inline fun I2OSP(value: Int, len: Int): ByteArray {
+    check(len == 2)
+    require(value in 0..0xffff)
+    return byteArrayOf(((value shr 8) and 0xff).toByte(), (value and 0xff).toByte())
+}
+
+private inline fun expand_message_xmd(digest: Digest): (msg: Sequence<ByteArray>, domain: ByteArray, lenInBytes: Int) -> Buffer {
+    check(digest.outputLength.bitSpacing == 0u) { "RFC9380 requirement: b mod 8 = 0 " }
+    check(digest.outputLength <= digest.inputBlockSize) { "RFC9380 requirement: b <= s" }
+    val sInBytes = digest.inputBlockSize.bytes.toInt()
+    val bInBytes = digest.outputLength.bytes.toInt()
+    return { msg, domain, lenInBytes ->
+        val ell = (lenInBytes.floorDiv(bInBytes) + if (lenInBytes.mod(bInBytes) != 0) 1 else 0).also {
+            require (it <= 255) { "RFC 9380 requirement: ell <= 255"}
+        }.toUByte()
+        require(lenInBytes <= 65535 && domain.size <= 255) { "RFC 9380 requirements" }
```

Comment:
the others explain the violated constraint, this one doesn't

Reply:
it should spell out the requirement, just as everywhere else

Reply:
done